MockJoinerTest.restart_multipleInParallel()

```java
java.lang.AssertionError: Cluster size is not correct. Expected: 5 Actual: 4
	at org.junit.Assert.fail(Assert.java:88)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSize(HazelcastTestSupport.java:870)
	at com.hazelcast.test.HazelcastTestSupport$12.run(HazelcastTestSupport.java:889)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1003)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:885)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:880)
	at com.hazelcast.test.mocknetwork.MockJoinerTest.assertClusterSizeEventually(MockJoinerTest.java:145)
	at com.hazelcast.test.mocknetwork.MockJoinerTest.restart_multipleInParallel(MockJoinerTest.java:139)
```
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/8738/testReport/junit/com.hazelcast.test.mocknetwork/MockJoinerTest/restart_multipleInParallel/

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterHeartbeatManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterHeartbeatManager.java
@@ -160,7 +160,13 @@ public class ClusterHeartbeatManager {
 
         try {
             if (!clusterService.isJoined()) {
-                logger.fine("Ignoring heartbeat of sender: " + senderMembersViewMetadata + ", because node is not joined!");
+                if (clusterService.getThisUuid().equals(receiverUuid)) {
+                    logger.fine("Ignoring heartbeat of sender: " + senderMembersViewMetadata + ", because node is not joined!");
+                } else {
+                    logger.fine("Sending explicit suspicion to " + senderAddress + " for heartbeat " + senderMembersViewMetadata
+                            + ", because this node has received an invalid heartbeat before it joins to the cluster");
+                    clusterService.sendExplicitSuspicion(senderMembersViewMetadata);
+                }
                 return;
             }
 

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java
@@ -427,6 +427,11 @@ public class MembershipManager {
 
         clusterServiceLock.lock();
         try {
+            if (!clusterService.isJoined()) {
+                logger.fine("Cannot handle suspect of " + suspectedMember + " because this node is not joined...");
+                return;
+            }
+
             ClusterJoinManager clusterJoinManager = clusterService.getClusterJoinManager();
             if (clusterService.isMaster() && !clusterJoinManager.isMastershipClaimInProgress()) {
                 removeMember(suspectedMember, reason, shouldCloseConn);
@@ -697,7 +702,7 @@ public class MembershipManager {
             futures.put(member.getAddress(), invokeFetchMembersViewOp(member.getAddress(), member.getUuid()));
         }
 
-        while (true) {
+        while (clusterService.isJoined()) {
             boolean done = true;
 
             for (Entry<Address, Future<MembersView>> e : new ArrayList<Entry<Address, Future<MembersView>>>(futures.entrySet())) {

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManagerCompat.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManagerCompat.java
@@ -67,6 +67,11 @@ public class MembershipManagerCompat {
 
         clusterServiceLock.lock();
         try {
+            if (!clusterService.isJoined()) {
+                logger.fine("Cannot remove " + deadAddress + " with uuid: " + uuid + " because this node is not joined...");
+                return;
+            }
+
             MemberImpl member = membershipManager.getMember(deadAddress);
             if (member == null || (uuid != null && !uuid.equals(member.getUuid()))) {
                 if (logger.isFineEnabled()) {

